apps/designer/app/shared/tree-utils.ts

```diff
@@ -0,0 +1,61 @@
+import type { Instance, Props, Styles } from "@webstudio-is/project-build";
+import { removeByMutable } from "./array-utils";
+
+const traverseInstances = (
+  instance: Instance,
+  cb: (child: Instance, parent: Instance) => void
+) => {
+  for (const child of instance.children) {
+    if (child.type === "text") {
+      continue;
+    }
+    if (child.type === "instance") {
+      cb(child, instance);
+      traverseInstances(child, cb);
+    }
+  }
+};
+
+export const deleteInstanceMutable = ({
+  rootInstance,
+  props,
+  styles,
+  deletedInstanceId,
+}: {
+  rootInstance: Instance;
+  props: Props;
+  styles: Styles;
+  deletedInstanceId: string;
+}) => {
+  const parentInstances = new Map<Instance["id"], Instance>();
+  const deletedInstances = new Set<Instance["id"]>();
+
+  traverseInstances(rootInstance, (child, instance) => {
```

Comment:
even in this case seems like an unnecessary function because all it does is this:

```suggestion
  for (const child of instance.children) {
    if (child.type !== "instance") {
      continue;
    }
};
```

Reply:
Not really, it's recursive function